technic/machines/HV/quarry.lua

```diff
@@ -184,402 +71,485 @@ local function can_dig_node(pos, node_name, owner, digger)
+	"field[4.3,3.1;2,1;offset_x;"..S("Offset X:")..";${offset_x}]"..
+	"field[6.3,3.1;2,1;offset_z;"..S("Offset Z:")..";${offset_z}]"
+
+if has_digilines then
+	base_formspec = base_formspec..
+	"field[4.3,4.2;4,1;channel;"..S("Digiline Channel:")..";${channel}]"
+end
 
-local function execute_dig(pos, node, meta, network)
-	local dig_pos = minetest.string_to_pos(meta:get_string("dig_pos"))
-	local quarry_dir = meta:get_int("quarry_dir")
-	if not dig_pos then
-		-- quarry has not hit ground yet
-		dig_pos = find_ground(pos, quarry_dir, meta)
+local function update_formspec(meta)
+	local fs = base_formspec
+	local status = S("Digging not started")
+	if meta:get_int("purge_on") == 1 then
+		status = S("Purging cache")
+	elseif meta:get_int("finished") == 1 then
+		status = S("Digging finished")
+	elseif meta:get_int("enabled") == 1 then
+		local diameter = meta:get_int("size") * 2 + 1
+		local num_steps = diameter * diameter
+		local y_level = math.floor(meta:get_int("step") / num_steps) + 1
+		status = S("Digging @1 m below machine", y_level)
+	end
+	if meta:get_int("enabled") == 1 then
+		fs = fs.."button[4,0.6;2,1;disable;"..S("Enabled").."]"
 	else
-		local owner = meta:get_string("owner")
-		local digger = get_or_create_fake_player(owner)
-		local dig_steps = meta:get_int("dig_steps")
-		local dig_index = meta:get_int("dig_index")
-		local t0 = minetest.get_us_time()
-		local us_used = 0
-		-- search for something to dig
-		while us_used < quarry_time_limit do
-			dig_pos, dig_index = get_dig_pos(pos, quarry_dir, dig_pos, dig_index, dig_steps, meta)
-			if not dig_pos then
-				-- finished digging
-				meta:set_int("finished", 1)
-				meta:set_int("purge_on", 1)
-				break
-			end
-			local dig_node = technic.get_or_load_node(dig_pos)
-			if can_dig_node(dig_pos, dig_node.name, owner, digger) then
-				-- found something to dig, dig it and stop searching
-				minetest.remove_node(dig_pos)
-				if quarry_dig_particles and network.lag < 35000 then
-					dig_particles(pos, dig_pos, dig_node)
-				end
-				local inv = meta:get_inventory()
-				local drops = minetest.get_node_drops(dig_node.name, "")
-				for _, dropped_item in ipairs(drops) do
-					local left = inv:add_item("cache", dropped_item)
-					while not left:is_empty() do
-						meta:set_int("purge_on", 1)
-						quarry_handle_purge(pos)
-						left = inv:add_item("cache", left)
-					end
-				end
-				local dug_nodes = meta:get_int("dug") + 1
-				meta:set_int("dug", dug_nodes)
-				if dug_nodes % 100 == 0 then
-					meta:set_int("purge_on", 1)
-				end
-				break
-			end
-			us_used = minetest.get_us_time() - t0
-		end
-		meta:set_int("dig_index", dig_index)
+		fs = fs.."button[4,0.6;2,1;enable;"..S("Disabled").."]"
 	end
-	if dig_pos then
-		meta:set_string("dig_pos", minetest.pos_to_string(dig_pos))
+	if quarry_dig_particles then
+		local selected = meta:get("particles") or "true"
+		fs = fs.."checkbox[0,3.6;particles;"..S("Enable Digging Particles")..";"..selected.."]"
```

Comment:
I think this shouldn't really be configurable per quarry but instead be global server configuration.

It is more environment effect for all players, not something that quarry owner should control.
If configurable per quarry then there probably should be another server configuration that enables or disables per quarry configuration...

My opinion: better to leave this to be server wide configuration and remove per quarry config + meta values so that effects as seen by players are same no matter what quarry node is removing nodes near them.

Reply:
Hmm, you make a good point... I think I will remove that.